docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,341 @@
+The processes that do not have an global iteration space index intersection with the provided regions simply skip this computation altogether.
+
+
+Usage and Impact
+----------------
+
+This is an optional feature, but should help nearly every application of GT4Py, since almost all partial differential equations at some point have a non-periodic boundary condition or require special treatment somewhere in the domain.
+
+
+Backward Compatibility
+----------------------
+
+This GDP aims to be fully backward-compatible.
+
+
+Detailed Description
+--------------------
+
+The intent of this GDP is to a specify horizontal bounds inside the stencil in a similar fashion to how vertical intervals are specified.
+However, the fact that there are multiple dimensions in the horizontal, halos, and domain decomposition imply the feature needs to be more complex than a simple index range used for the vertical.
+
+As introduced above, we propose adding a new gtscript `region()` specifier that applies the stencil on a region of the space when a non-zero intersection with the global region specified occurs.
+The idea behind using the region objects is that they can incorporate both the iteration space a process sees and the global partitioning that an application must handle, all while hiding the complexity at the stencil level for model developers.
+In this proposal, the application remains in control of the partitioning data so **the partitioner is entirely transparent to both GT4Py and its backend.**
+
+To explain how this works, consider part of the example above with a domain of size N by N by K points, decomposed by three processes using vertical strips of the domain:
+
+.. math:: \mathrm{proc}\, 0: [0,n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 1: [n,2n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 2: [2n,N) \times [0,N) \times [0,K).
+
+.. code-block:: python
+
+    with region(WestEdge(0, 1), EastEdge(0, 1)), interval(...):
+
+This part of the example neatly indicates that it is intended to be run on both the east and west edges of horizontal plane for each `k`, with an index strip `1` wide.
+The construction and evaluation of these application-specific objects, `WestEdge(0,1)` and `EastEdge(0,1)`, are done before code generation, and can call into user-provided code.
+These objects have attributes `local_origin` and `extent` which define the iteration space in the horizontal plane `i` and `j`.
+The union of the boxes defined by each of these objects is used for iteration.
+
+When GT4Py asks for the iteration space from the region objects, it sees different `local_origin` and `extent` on each process using the local iteration space indices:
+
+.. code-block:: python
+
+    proc 0:    local_origin=(0, 0), extent=(1, N)
+    proc 1:    local_origin=(0, 0), extent=(0, 0)
+    proc 2:    local_origin=(n-1, 0), extent=(1, N)
```

Comment:
this would lead to separate code on separate processors, right?

Reply:
Yes. Code generation could keep the loop on proc 1, but the bounds would be such that it doesn't execute.

At some point we need different code executed, right? We can also include information about the entire loop if you think that would help.